### DISCUSSION TITLE
rhel7.9 Failed to load Keytar module: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.21'

### DISCUSSION THREAD

**Discussion by couckearthur** [zowe/zowe-cli#1587] [Q&A] [answered]
For a CI/CD project we are installing Zowe CLI on rhel7.9 server. The goal is to use it in a Jenkins pipeline. 

When creating a profile we are receiving the following error: 

-----------------------------------------------------------------------
"Failed to load Keytar module: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.21'"
-----------------------------------------------------------------------

When listing out /usr/lib64/libstdc++.so.6.0.19 I see that we are indeed missing version 3.4.21

-----------------------------------------------------------------------
+ strings /usr/lib64/libstdc++.so.6.0.19
+ grep GLIBCXX
GLIBCXX_3.4
GLIBCXX_3.4.1
GLIBCXX_3.4.2
GLIBCXX_3.4.3
GLIBCXX_3.4.4
GLIBCXX_3.4.5
GLIBCXX_3.4.6
GLIBCXX_3.4.7
GLIBCXX_3.4.8
GLIBCXX_3.4.9
GLIBCXX_3.4.10
GLIBCXX_3.4.11
GLIBCXX_3.4.12
GLIBCXX_3.4.13
GLIBCXX_3.4.14
GLIBCXX_3.4.15
GLIBCXX_3.4.16
GLIBCXX_3.4.17
GLIBCXX_3.4.18
GLIBCXX_3.4.19
GLIBCXX_DEBUG_MESSAGE_LENGTH
-----------------------------------------------------------------------

The Linux sysadmin seems to be struggling with it: 
"We tried searching this rpm from local and external repos but could find right one. May be support team could help us on rhel7.9."

Anyone bumped into this? Suggestions?

**Comment by zFernand0** [comment, marked as answer]
This sounds related to a recent roadblock we came across with CentOS 7
With @t1m0thyj 's help, we came up with this little script.

```shell
curl -fsSL https://rpm.nodesource.com/setup_14.x | bash - && yum install -y nodejs
yum install -y gcc-c++ libsecret-devel make python3 dbus-x11 gnome-keyring
npm install -g @zowe/cli --foreground-scripts --verbose
cd $(npm prefix -g)/lib/node_modules/@zowe/cli/node_modules/keytar && npm run build
mkdir -p $HOME/.local/share/keyrings
export $(dbus-launch) && echo root | gnome-keyring-daemon -r --unlock --components=secrets
```

**Notes:**
- Some commands may require `sudo` access.
- `--foreground-scripts --verbose` are there to help diagnose other issues that may occur.
- If `zowe config init --global-config` is not able to store credentials after these commands, consider unlocking the keyring (last command) with the `-f, --foreground` option to further diagnose other issues.
  - You may need to `Ctrl + C` out of the `gnome-keyring-daemon --foreground -r --unlock --components=secrets` since it takes control of the terminal.

**Reply by couckearthur** [reply]
@zFernand0 Thank you for the quick response, the way you described solved the problem.

**Comment by zFernand0** [comment]
> This sounds related to a recent roadblock we came across with CentOS 7 With @t1m0thyj 's help, we came up with this little script.
> 
> ```shell
> curl -fsSL https://rpm.nodesource.com/setup_14.x | bash - && yum install -y nodejs
> yum install -y gcc-c++ libsecret-devel make python3 dbus-x11 gnome-keyring
> npm install -g @zowe/cli --foreground-scripts --verbose
> cd $(npm prefix -g)/lib/node_modules/@zowe/cli/node_modules/keytar && npm run build
> mkdir -p $HOME/.local/share/keyrings
> export $(dbus-launch) && echo root | gnome-keyring-daemon -r --unlock --components=secrets
> ```

These commands are slightly out of date.

You may want to use node 20
- `curl -fsSL https://rpm.nodesource.com/setup_20.x | bash - && yum install -y nodejs`

Also we don't have keytar as a dependency anymore, so you will have to try building the secrets SDK instead.
- `cd $(npm prefix -g)/lib/node_modules/@zowe/cli/node_modules/@zowe/secrets-for-zowe-sdk && npm run build`

This may error out because of possible `npm install` issues.

---

UPDATE: 
```shell
# Install Node.js and npm
curl -fsSL https://rpm.nodesource.com/setup_20.x | bash - && sudo yum install -y nodejs npm

# Install other required dependendencies
sudo yum install -y gcc-c++ libsecret-devel make python3 dbus-x11 gnome-keyring wget

# Install the Zowe CLI
npm install -g @zowe/cli --foreground-scripts --verbose

# Replace keyring binary
cd $(npm prefix -g)/lib/node_modules/@zowe/cli/node_modules/@zowe/secrets-for-zowe-sdk/prebuilds
sudo rm keyring.linux-x64-gnu.node
sudo wget https://github.com/zFernand0/cics-for-zowe-client/releases/download/v0.0.1/keyring.linux-x64-gnu.node

# Unlock the keyring
# You should be able to follow the Zowe documentation.
# https://docs.zowe.org/stable/user-guide/cli-configure-scs-on-headless-linux-os

# If that does not work you can try the commands below to setup a known dbus_address
# export DBUS_SESSION_BUS_ADDRESS=unix:path=$XDG_RUNTIME_DIR/bus
# if [ ! -e "$bus_file_path" ]; then
#     /usr/bin/dbus-daemon --session --address=$DBUS_SESSION_BUS_ADDRESS --nofork --nopidfile &
# fi
# echo <sudo-password> | gnome-keyring-daemon -r --unlock --components=secrets
# export GNOME_KEYRING_CONTROL=$XDG_RUNTIME_DIR/keyring
```
